edgeless_node/src/resources/ollama.rs

```diff
@@ -0,0 +1,344 @@
+                    }
+                };
+
+                let (reply_sender, reply_receiver) =
+                    tokio::sync::oneshot::channel::<anyhow::Result<(edgeless_api::function_instance::InstanceId, String)>>();
+                let _ = sender
+                    .send(OllamaCommand::Chat(
+                        model_name.clone(),
+                        history_id.clone(),
+                        message_data,
+                        instance_id.function_id.clone(),
+                        reply_sender,
+                    ))
+                    .await;
+
+                match reply_receiver.await {
+                    Ok(response) => match response {
+                        Ok((target, response)) => {
+                            let _ = dataplane_handle.send(target, response).await;
+                        }
+                        Err(err) => {
+                            log::warn!("Error from ollama: {}", err)
+                        }
+                    },
+                    Err(err) => {
+                        log::warn!("Communication error with ollama resource provider: {}", err)
+                    }
+                };
+
+                if need_reply {
+                    dataplane_handle
+                        .reply(source_id, channel_id, edgeless_dataplane::core::CallRet::Reply("".to_string()))
+                        .await;
+                }
+            }
+        });
+
+        Ok(Self { join_handle: handle })
+    }
+}
+
+impl OllamaResourceProvider {
+    /// Create an Ollama resource provider:
+    ///
+    /// - `dataplane_provider`: handle to the EDGELESS data plane
+    /// - `resource_provider_id`: identifier of this resource provider,
+    ///    also containing the identifier of the node hosting it
+    /// - `ollama_host`: address of the ollama server
+    /// - `ollama_port`: port number of the ollama server
+    /// - `ollama_messages_number_limit`: maximum number of messages per
+    ///    chat conversation
+    pub async fn new(
+        dataplane_provider: edgeless_dataplane::handle::DataplaneProvider,
+        resource_provider_id: edgeless_api::function_instance::InstanceId,
+        ollama_host: &str,
+        ollama_port: u16,
+        ollama_messages_number_limit: u16,
+    ) -> Self {
+        // Create a channel for:
+        // - single receiver: the loop in the task below
+        // - multiple senders: the resource instances that will be created
+        //   at run-time
+        let (sender, receiver) = futures::channel::mpsc::unbounded();
+        let mut receiver: futures::channel::mpsc::UnboundedReceiver<OllamaCommand> = receiver;
+
+        // Create a new instance of the ollama connector.
+        let mut ollama = ollama_rs::Ollama::new_with_history(format!("http://{}", ollama_host), ollama_port, ollama_messages_number_limit);
+
+        let _handle = tokio::spawn(async move {
+            let mut targets = std::collections::HashMap::new();
```

Comment:
I don't quite get why you need this / why the patch requests are forwarded to the provider (also see comment above).

Reply:
It was done on purpose. Other designs are possible. See my response to the previous comment.

Reply:
Even if the above applies, the response is sent back to the single resource using the oneshot.
So only that needs to know about the target (and it is also using two strategies in that).

Reply:
Yes, that's correct. However, the mapping between the resource_id and the target is dynamic and unknown to the `OllamaResource`. That's why I keep the mapping between all resource_ids and targets in the resource _provider_ and send the latest target (following the last Patch command) to the `OllamaResource`.

So, `targets` is not redundant, as implied by your comment.

On the other hand, I understand this can be rewritten in several ways 🤷‍♂️